EXILED/Exiled.API/Features/Npc.cs

```diff
@@ -131,55 +133,69 @@ public Npc(GameObject gameObject)
             {
-                Log.Debug($"Ignore: {e}");
+                Log.Warn($"{Assembly.GetCallingAssembly().GetName().Name} tried to spawn an NPC with a duplicate PlayerID. Using auto-incremented ID instead to avoid an ID clash.");
+                id = new RecyclablePlayerId(true).Value;
             }
 
-            if (RecyclablePlayerId.FreeIds.Contains(id))
+            try
             {
-                RecyclablePlayerId.FreeIds.RemoveFromQueue(id);
+                if (userId == PlayerAuthenticationManager.DedicatedId)
+                {
+                    npc.ReferenceHub.authManager.SyncedUserId = userId;
+                    try
+                    {
+                        npc.ReferenceHub.authManager.InstanceMode = ClientInstanceMode.DedicatedServer;
+                    }
+                    catch (Exception e)
+                    {
+                        Log.Debug($"Ignore: {e.Message}");
+                    }
+                }
+                else
+                {
+                    npc.ReferenceHub.authManager.InstanceMode = ClientInstanceMode.Unverified;
+                    npc.ReferenceHub.authManager._privUserId = userId == string.Empty ? $"Dummy@localhost" : userId;
+                }
             }
-            else if (RecyclablePlayerId._autoIncrement >= id)
+            catch (Exception e)
             {
-                RecyclablePlayerId._autoIncrement = id = RecyclablePlayerId._autoIncrement + 1;
+                Log.Debug($"Ignore: {e.Message}");
             }
 
-            FakeConnection fakeConnection = new(id);
-            NetworkServer.AddPlayerForConnection(fakeConnection, newObject);
             try
             {
-                npc.ReferenceHub.authManager.UserId = string.IsNullOrEmpty(userId) ? $"Dummy@localhost" : userId;
+                npc.ReferenceHub.roleManager.InitializeNewRole(RoleTypeId.None, RoleChangeReason.None);
             }
             catch (Exception e)
             {
-                Log.Debug($"Ignore: {e}");
+                Log.Debug($"Ignore: {e.Message}");
             }
 
+            FakeConnection fakeConnection = new(id);
+            NetworkServer.AddPlayerForConnection(fakeConnection, newObject);
+
             npc.ReferenceHub.nicknameSync.Network_myNickSync = name;
             Dictionary.Add(newObject, npc);
 
-            Timing.CallDelayed(
-                0.3f,
-                () =>
-                {
-                    npc.Role.Set(role, SpawnReason.RoundStart, position is null ? RoleSpawnFlags.All : RoleSpawnFlags.AssignInventory);
-                });
+            Timing.CallDelayed(0.5f, () =>
+            {
+                npc.Role.Set(role, SpawnReason.RoundStart, position is null ? RoleSpawnFlags.All : RoleSpawnFlags.AssignInventory);
+            });
 
             if (position is not null)
                 Timing.CallDelayed(0.5f, () => npc.Position = position.Value);
+
```

Comment:
If you put the same delay then you can combine them

Reply:
No, because if you set the role and then change the position and it has the SpawnPoint flag it will be moved to spawn and ignore the position variable, that's why it has 2 delays.

Reply:
`if(position is not null) npc.Position = position.Value`
after the role set not working ?

Reply:
> No, because if you set the role and then change the position and it has the SpawnPoint flag it will be moved to spawn and ignore the position variable, that's why it has 2 delays.

yeah i don't see where it's problem you can just put the check inside your delayed call